### ISSUE TITLE
`org-journal-previous-entry` and `org-journal-next-entry` opens wrong file

### ISSUE THREAD

**Issue by vitorqb** [bastibe/org-journal#272]
Hi!

### Problem 
For some reason when I call `org-journal-previous-entry`, org-journal always opens the oldest journal entry of all instead of the previous entry. When I call `org-journal-next-entry`, it always opens the last journal entry instead of the next one.

### Version
```
(define-package "org-journal" "20200702.519" "a simple org-mode based journaling mode" '((emacs "25.1") (org "9.1")) :commit "b4b3b2d9854ec9ba98c16098979d2b6b40f54180" :authors '(("Bastian Bechtold") ("Christian Schwarzgruber")) :maintainer '("Bastian Bechtold") :url "http://github.com/bastibe/org-journal")
```

```
GNU Emacs 26.3 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.20) of 2020-05-19
```

### Configuration
```
(custom-set-variables
 '(org-journal-carryover-items "TODO=\"TODO\"|TODO=\"WAIT\"")
 '(org-journal-tag-alist '(("dev" . ?d) ("meeting" . ?m) ("short" . ?s) ("learning" . ?l)))
 '(org-journal-date-format "%A, %Y%m%d")
 '(org-journal-dir "/home/vitor/gpm-org/journal/"))
```

### Investigation
I see that `org-journal-open-entry` is calling `org-journal-list-dates`, and then we have this line:

```
(when prev
      (setq dates (reverse dates)))
```

If I print `dates` __after__ this line, it's sorted from smaller to biggest:

```
((6 3 2019) (6 4 2019) (6 5 2019) ... (7 3 2020) (7 6 2020) (7 7 2020))
```

I have a feeling that this should be the other way around. Actually if I change the line for 
```
(when (not prev)
      (setq dates (reverse dates)))
```

it all works. So I'm assuming that `org-journal-list-dates` was supposed to return the dates in a specific order, but for some reason in this journal it seems to be listing the dates in the reverse order.

Let me know if I can help debugging it further!

__EDIT__  
Added missing configuration.
Added emacs version.

**Comment by bastibe** [comment]
This sure sounds like a bug. Did this behavior change recently? It works correctly in the stable version, as far as I can tell.

**Comment by vitorqb** [comment]
@bastibe I actually use more than one journal, and this bug only happens for a specific journal and on some machines. I tried both the stable and dev versions and both had the same issue.

I was wondering: it seems that there is an implicit assumption that `org-journal-list-dates` returns the dates in a specific order, but I don't see any explicit sorting happening in it, and it ends with `(hash-table-values org-journal-dates)`. Could it be that the hashing algorithim used by different platform or emacs versions, or even the strategy for different file names and dates, could be causing this issue? Would it make sense to put an explicit sorting on `org-journal-list-dates`?

**Comment by bastibe** [comment]
I stumbled upon that issue as well. But I am not sure, honestly. Most of the development of org-journal has been taken over by @cslux. I would say, let's wait for him to comment on this. Org-journal is really his project these days.

**Comment by cslux** [comment]
Sorry for letting you guys wait so long.  I guess this related to #264. I'll need to reproduce this may self, preferable with a test. Than I'll fix it. 

> it seems that there is an implicit assumption that `org-journal-list-dates` returns the dates in a specific order, but I don't see any explicit sorting happening in it, and it ends with `(hash-table-values org-journal-dates)`

Indeed, we rely on a specific date order.

> Could it be that the hashing algorithim used by different platform or emacs versions, or even the strategy for different file names and dates, could be causing this issue?

I don't know, but can very likely be the case.

> Would it make sense to put an explicit sorting on `org-journal-list-dates`?

That sounds good! But still, it would be nice if we find out what was causing it.

**Comment by cslux** [comment]
> If I print `dates` **after** this line, it's sorted from smaller to biggest:

That's correct. 

Those are my journal entries.
```
((11 5 2015) (11 9 2015) (5 1 2017) (2 22 2018) (5 17 2018) (5 18 2018) (6 24 2018) (6 26 2018) (7 3 2018) (7 18 2018) (7 27 2018) (1 13 2019) (3 23 2019) (3 27 2019) (4 13 2019) (4 16 2019) (4 17 2019) (4 21 2019) (4 23 2019) (5 5 2019) (5 18 2019) (6 1 2019) (6 16 2019) (6 30 2019) (7 14 2019) (7 28 2019) (8 11 2019) (8 25 2019) (9 1 2019) (9 8 2019) (9 15 2019) (9 22 2019) (9 30 2019) (10 6 2019) (10 10 2019) (10 13 2019) (10 15 2019) (10 20 2019) (10 25 2019) (10 27 2019) (10 31 2019) (11 3 2019) (11 10 2019) (11 18 2019) (11 23 2019) (12 1 2019) (12 8 2019) (12 15 2019) (12 22 2019) (12 29 2019) (1 5 2020) (1 12 2020) (1 19 2020) (1 26 2020) (2 2 2020) (2 9 2020) (2 16 2020) (2 23 2020) (2 27 2020) (3 1 2020) (3 8 2020) (3 15 2020) (3 22 2020) (3 29 2020) (4 5 2020) (4 13 2020) (4 19 2020) (4 26 2020) (5 3 2020) (5 17 2020) (5 25 2020) (5 31 2020) (6 7 2020) (6 8 2020) (6 14 2020) (6 21 2020) (7 5 2020) (7 6 2020) (7 13 2020))
```

> (when (not prev)
>       (setq dates (reverse dates)))

That's ok, if the return value of `org-journal-list-dates` is sorted correctly; oldest -> newest

I have generated daily journal files from my yearly journal files, but so far I wasn't able to reproduce it.

**Comment by vitorqb** [comment]
@cslux thanks for taking a look into that.

I must say the behavior seems quite erratic, and I can only reproduce it in a single computer (that I use for work, so it's hard to find the time to debug it). But the exact same files on my personal pc (exact same as in it's a shared in a cloud synchronization services without any exception rule) work just fine. I'm running different emacs versions though (the bug happens on 26.3, and my personal pc is in 27.0.91).

I still think the issue is that we can not assume `hash-table-values`  will return the values of the hash table in any predictable order. I tried to find some information about the hashing algorithim in emacs but I couldn't find anything useful. But the documentation for `hash-table-values` does not guarantee it will be sorted in any way, and that's usually the case for hashing maps in most languages.

But since you see the same ordering in your journal, I'll try to find some time tomorrow to dig more into this and see if I find anything else. I'll post it here if I do :)

**Comment by vitorqb** [comment]
@cslux So, that's the flow on the pc with the bug, when I'm in 2020-07-15 and I'm opening the **previous** entry:

1. `(org-journal-list-dates)` returns dates ordered from **bigger->smaller** (newest->oldest) at line 10.
2. `(when prev (setq dates (reverse dates))` at lines 22/23 resorts it into **smaller->bigger** (oldest->newest).
3. When we reach the `while` at line 24, the first date on dates is lower (older) than the `calendar-date` (e.g. in my debugging, `(car dates)` is (6 3 2019) and `calendar-date` is (7 15 2020)).
4. The `or` condition on line 25 is not met since both `(calendar-date-compare (list calendar-date) dates)` and `(calendar-date-equal (car dates) calendar-date))` return nil
5. Line 31 is reached, the `(car dates)` returns (6 3 2019), and that's the file that ends up being opened.

```elisp
1  >(defun org-journal-open-entry (&optional prev no-select)
2  >  "Open journal entry.
3  >If no next/previous entry was found print MSG."
4  >  (let ((calendar-date (if (org-journal-daily-p)
5  >                           (org-journal-file-name->calendar-date (file-truename (buffer-file-name)))
6  >                         (while (org-up-heading-safe))
7  >                         (org-journal-entry-date->calendar-date)))
8  >        (view-mode-p view-mode)
9  >        (dates (org-journal-list-dates)))
10 >    (unless (member calendar-date dates)
11 >      ;; Insert calendar-date into dates list keeping it in order.
12 >      (setq dates (cl-loop
13 >                     for date in dates
14 >                     while (calendar-date-compare (list date) (list calendar-date))
15 >                     collect date into result and count t into cnt
16 >                     finally return (if result
17 >                                        ;; Front
18 >                                        `(,@result ,calendar-date)
19 >                                      ;; Somewhere enbetween or end of dates
20 >                                      `(,calendar-date ,@result ,@(nthcdr cnt dates))))))
21 >    ;; Reverse list for previous search.
22 >    (when prev
23 >      (setq dates (reverse dates)))
24 >    (while (and dates (car dates)
25 >                (or (if prev
26 >                        (calendar-date-compare (list calendar-date) dates)
27 >                      (calendar-date-compare dates (list calendar-date)))
28 >                    (calendar-date-equal (car dates) calendar-date)))
29 >      (setq dates (cdr dates)))
30 >    (if (and dates (car dates))
31 >        (let* ((date (car dates))
32 >               (time (org-journal-calendar-date->time date))
33 >               (filename (org-journal-get-entry-path time)))
34 >          (if (get-file-buffer filename)
35 >              (progn
36 >                (if no-select
37 >                    (set-buffer (get-file-buffer filename))
38 >                  (switch-to-buffer (get-file-buffer filename)))
39 >                (setq org-journal--kill-buffer nil))
40 >            (push (if no-select
41 >                      (set-buffer (find-file-noselect filename))
42 >                    (find-file filename))
43 >                  org-journal--kill-buffer))
44 >          (org-journal-goto-entry date)
45 >          (view-mode (if view-mode-p 1 -1))
46 >          t)
47 >      nil)))
```

**Comment by cslux** [comment]
@vitorqb Thanks for helping. Interesting, `org-journal-list-dates` returns the dates sorted from oldest -> newest for me, which would be the expected order, not newest -> oldest. Which means `(hash-table-values org-journal-dates)` returns the dates for you in the order newest -> oldest, instead of oldest -> newest. I'll add something to insure we always get the dates sorted from oldest -> newest.

**Comment by cslux** [comment]
Should be fixed with commit cf1017b7e6e70d73ab691835e758b15a1ac8b3b0. Thanks for your help @vitorqb.
